### ISSUE TITLE
End interactions prematurely

### ISSUE THREAD

**Issue by sebbonnet** [plannigan/columbo#486]
Would you be open to add a new type of interaction to end the interactions prematurely?

At the moment, the user has to go through all interactions even when there are no more branches available.
There are situations where it would be useful to gracefully terminate the interactions before the end, for instance when the question is based on "dynamic values" that are collected from an external source - e.g. an API call. If that API call is broken or doesn't return anything, there are no choice possible for the user to select and so the rest of the questions might not be required.
The `should_ask` attribute mitigates that problem, but requires every subsequent questions to use that flag which may be cumbersome and makes it hard to follow the flow of questions.
 
Here is a use case to illustrate the problem:
```
        columbo.Choice(
            "favourite_car",
            "What is your favourite car brand?",
            options=get_cars, # makes an API call to retrieve a list of cars
            default=get_first_car, # gets the first car from the list
            should_ask=some_cars_were_found,
        ),
        columbo.Acknowledge(
            "No cars were found! Please come back later!",            
            should_ask=lambda answers: not some_cars_were_found(answers),
        ),
        columbo.Choice(
            "favourite_car_colour",
            f"What is your favourite color for car {answers['favourite_car']}",
            options=get_colors  # makes an API call to retrieve a list of colors
            default=get_any_color, # randomise the default color 
            should_ask=lambda answers: some_cars_were_found(answers) and some_colors_were_found(answers),
        ),
        columbo.Choice(
            "favourite_car_model",
            f"Choose a model for your {answers['favourite_car']}",
            options=get_models
            default=get_any_model,
            should_ask=lambda answers: some_cars_were_found(answers) and some_colors_were_found(answers) and some_models_were_found(answers),
        ),
```

If the `get_cars` API call doesn't return anything, then the rest of the questions are no longer relevant.
You can imagine that if you have a lot more questions based on cars, the `should_ask` property is going to look horrible.

Perhaps we could introduce a new interaction that ends the questionnaire early - e.g. having an `EndStatement` that behaves like an `Acknowledge` but breaks the interactions loop once executed.
Please let me know if this is something you would consider, so I can submit a proposal PR.

**Comment by plannigan** [comment]
It is an interesting idea that I would need to think more about. I acknowledge that `should_ask` could get unwieldy in the situation you described.

Some initial thoughts:
* If `get_cars()` produces an empty collection, a `ValueError` will be raised it doesn't provide the user any options. (This might not be relevant as it could be that the example was overly simplified and the actual situation doesn't have this issue.)
* Alternate options:
  * Instead of passing `get_cars()` to `options`, `get_cars()` could be called before constructing the interaction list. This would allow the code to check the result and then dynamically construct the list of interactions. This would avoid needing to complicate `columbo` based on specific behavior of external components.
  * If you know the Nth interaction is a point where it might make sense to stop, the interactions could be partitioned into two lists. `get_answers()` can be called with the fist list. Then the program checks the results of the Nth interaction. If it doesn't trigger a stop, `get_answers()` can be called again with the second list and the first results in the `answers` argument.
* The way the documentation is currently written, it may give users the impression that the only approach is to provide a single static list of interaction. I should add a section discussing the alternate options I just mentioned to make it clearer that that is not a requirement.
* You suggested an `EndStatement` interaction. My first thought would have been to change `ask()` to be `ask(...) -> Union[Answer, Stop]`. Not sure which would work best as my idea would make things more complicated for users that are just using the interactions instead of using `get_answers()`.
* How would this work with the CLI parsing? Does it get ignored? Or does it restrict which parameters are added to the parser?

Similar to #488, I'm not strictly opposed to the idea. But I would like to see a more fleshed out proposal before deciding if it makes sense to be added.

**Comment by sebbonnet** [comment]
> Instead of passing get_cars() to options, get_cars() could be called before constructing the interaction list. This would allow the code to check the result and then dynamically construct the list of interactions. This would avoid needing to complicate columbo based on specific behavior of external components.

This makes a lot of sense. 
Partitioning the user interactions in multiple `get_answers()` calls makes the code easier to read and the error handling much nicer. It also allows to get intermediate details externally in between user interactions.

I'm happy to close this.
Thanks for the feedback!
